### PR TITLE
Added Finnish Personal Identity Code Recognizer.

### PR DESCRIPTION
**Pull request by honderr**
## Change Description
- Added Pattern Recognizer for the Finnish Personal Identity Code (Henkilötunnus).
- Added validation logic for the Finnish Personal Identity Code.
- Added tests for different valid and invalid Finnish Personal Identity Codes.

Describe your changes

## Issue reference

This PR fixes issue #1346 

## Checklist

- [x] I have reviewed the [contribution guidelines](https://github.com/microsoft/presidio/blob/main/CONTRIBUTING.md)
- [x] I have signed the CLA (if required)
- [x] My code includes unit tests
- [x] All unit tests and lint checks pass locally
- [x] My PR contains documentation updates / additions if required

### CODE REVIEW

**Comment by honderr** [comment]
@microsoft-github-policy-service agree

File: presidio-analyzer/tests/test_fi_personal_identity_code_recognizer.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        (
+            "020504B904H",
+            1,
+            ((0, 11),),
+        ),
+        # invalid Personal Identity Codes scores
+        ("111111-111A", 0, ()),
+        ("111111+110G", 0, ()),
+        ("012245A110G", 0, ()),
+        ("010324A110G", 0, ()),
```
Comment (omri374) [inline presidio-analyzer/tests/test_fi_personal_identity_code_recognizer.py:132]: Please add a sample with the personal identity code within a sentence (e.g "my personal identity code, ABCD, thank you".

Comment (honderr) [reply presidio-analyzer/tests/test_fi_personal_identity_code_recognizer.py:132]: Thanks a lot for your feedback @omri374 

I would kindly ask for one more piece of advice, before making the changes. Following the example of other pattern recognizers, I used word boudaries `\b` in my regular expression.

In real-life cases, we might have phrases like "my personal id is111111-111A" which would not be caught. This is, in my experience, a common issues especially in chat contexts.

Do you recommend:
1. Keeping the solution as is.
2. Removing word boundaries and reducing the score of the recognizer.
3. Adding another (very weak e.g. 0.1 score) Pattern to the list without the word boundaries.

Thanks!

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    :param patterns: List of patterns to be used by this recognizer
+    :param context: List of context words to increase confidence in detection
+    :param supported_language: Language this recognizer supports
+    :param supported_entity: The entity this recognizer can detect
+    """
+
+    PATTERNS = [
+        Pattern(
+            "Finnish Personal Identity Code (Medium)",
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py:19]: Another point, if ok- please add a reference to the logic (a government website if possible) for future users to see where this logic is coming from.

Comment (honderr) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py:19]: @omri374 thanks. I have now made the changes.

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            supported_language=supported_language,
+        )
+
+    def validate_result(self, pattern_text: str) -> Optional[bool]:
+        """Validate the pattern by using the control character."""
+
+        # More information on the validation logic from:
+        # https://dvv.fi/en/personal-identity-code
+        # Under "How is the control character for a personal identity code calculated?".
+        date_part = pattern_text[0:6]
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py:53]: Please check for len prior to running this. If in the future the pattern changes and we pass shorter strings, it will raise an exception and stop the whole process.

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py (hunk 2 above)
Comment (omri374) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py:19]: Thanks!

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py (hunk 3 above)
Comment (honderr) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py:53]: Thanks a lot for your feedback @omri374 . In my understanding, the regular expression is already taking care of that?

My understanding is that, when entering `validate_result`, we have already matched `(\d{6})([+-ABCDEFYXWVU])(\d{3})([0123456789ABCDEFHJKLMNPRSTUVWXY])` and therefore, we can be sure that the string we mainpulate (`pattern_text`) is already following the right format (`pattern_text[0:6]` are all digits). 

In other words, I understand that the regular expression is already checking the length for us? Is my understanding correct or did I misunderstand your previous comment?

Comment (omri374) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py:53]: Yes, you are correct. I wasn't clear enough on my comment. Regex patterns sometimes change, and adding another check for len here is useful as we might unexpectedly break something if the pattern changes but the checksum doesn't. An exception here would not only affect the users of this recognizer, but all users of the tool. Hope this makes sense.

Comment (honderr) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py:53]: Thanks @omri374. Would something like this suffice?

```
if len(pattern_text) != 11:
    return False
```

Then, the, `datetime.strptime(date_part, "%d%m%y")` will take care of validating the `date_part`.

While I understand that we would like to avoid exceptions in this context, I must admit that I fail to think of scenarios where we would change the pattern without changing the checksum and, the tests would still pass?

Comment (honderr) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/fi_personal_identity_code_recognizer.py:56]: Thanks a lot for your reply.

I have now added the length check.

**Review by omri374 — APPROVED** [review]

**Comment by omri374** [comment]
Thanks for the great contribution! Merged.
